File: 프로그래머스/lv2/92341. 주차 요금 계산/주차 요금 계산.py
from collections import defaultdict
import math

def solution(fees, records):
    d = defaultdict(list)
    result = defaultdict(int)
    order = []
    # 정보 가공
    for record in records:
        time, number, io = record.split()
        if number not in d:
            order.append(number)
        d[number].append((time, io))
    
    # 마지막이 IN으로 끝나면 OUT을 추가해줌
    for number in d:
        if d[number][-1][1] == 'IN':
            d[number].append(('23:59', 'OUT'))
    
    # 시간계산 해서 result에 저장
    for number in d:
        target = d[number]
        for i in range(0, len(target), 2):
            out_time = int(target[i+1][0][0:2]) * 60 + int(target[i+1][0][3:])
            in_time = int(target[i][0][0:2]) * 60 + int(target[i][0][3:])
            result[number] += (out_time - in_time)
    
    # 요금 계산
    default_time = fees[0]
    default_fee = fees[1]
    unit_time = fees[2]
    unit_fee = fees[3]

    answer = []
    order.sort()
    for number in order:
        time = result[number]
        if time <= default_time:
            answer.append(default_fee)
            continue
        answer.append(default_fee + math.ceil((time - default_time) / unit_time) * unit_fee)
            
    
    return answer

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.js
function solution(fees, records) {
    let recordObject = {}
    let times = {}
    let feeArray = [];
    
    // 1. 데이터 가공
    records.forEach(element => {
        const [time, number, type] = element.split(' ')
        if (recordObject[number]) recordObject[number].push(time)
        else recordObject[number] = [time]
    })
    // 2. 만약 출차 시간이 없다면 23:59 추가
    for (const record in recordObject) {
        if (recordObject[record].length % 2 !== 0) recordObject[record].push('23:59')
    }
    // 3. 총 주차시간 계산(분 단위로 계산해야 함)
    for (const key in recordObject) {
        const record = recordObject[key]
        for (let i=0; i<record.length; i+=2) {
            const [inHour, inMinutes] = record[i].split(':').map((v) => +v)
            const [outHour, outMinutes] = record[i+1].split(':').map((v) => +v)
            const inTime = new Date(`2024/07/05 ${inHour}:${inMinutes}:00`)
            const outTime = new Date(`2024/07/05 ${outHour}:${outMinutes}:00`)
            const diffMinutes = (outTime.getTime() - inTime.getTime()) / (60 * 1000)
            if (times[key]) times[key] += diffMinutes
            else times[key] = diffMinutes
        }
    }
    // 4. 주차 요금 계산 : 기본 시간(분), 기본 요금(원), 단위 시간(분), 단위 요금(원)
    for (const key in times) {
        const time = times[key]
        let fee = 0
        if (time <= fees[0]) fee = fees[1]
        else {
            fee = fees[1] + Math.ceil((time - fees[0]) / fees[2]) * fees[3]
        }
        feeArray.push([key, fee])
    }
    // 5. 차량번호 오름차순 정렬
    feeArray.sort((a, b) => +a[0] - +b[0])
    return feeArray.map((v) => v[1]);
}